06x-new-kafka-roller.md

```diff
@@ -0,0 +1,220 @@
+ - <i>numReconfigs</i>: The value is incremented each time the node has been retried by the roller.
+ - <i>lastTransitionTime</i>: System.currentTimeMillis of last observed state transition.
+
+ <b>States</b>
+ | State        | Description |
+ | :-------------------- | :---------- |
+ | UNKNOWN               | The initial state when creating `Context` for a node. We expect to transition from this state fairly quickly after creating the context for nodes.  |
+ | NOT_RUNNING           | Node is not running (Kafka process is not running) |
+ | NOT_READY             | Node is running but not ready to serve requests (broker state < 2 OR == 127) |
+ | RESTARTED             | After successful `kubectl delete pod`. |
+ | RECONFIGURED          | After successful Kafka node config update via Admin client. |
+ | RECOVERING            | Node has started but is in log recovery (broker state == 2). |
+ | SERVING               | Node is in running state and ready to serve requests (broker state >= 3 AND != 127). |
+ | LEADING_ALL_PREFERRED | Node is in running state and leading all preferred replicas. |
+
+2. The existing predicate function will be called for each of the nodes and those for which the function returns a non-empty list of reasons will be restarted.
+
+3. Group the nodes into the following categories based on information collected from the abstracted sources:
+   - `RESTART_FIRST` - Nodes that have `NOT_READY` or `NOT_RUNNING` state in their contexts. The group will also include nodes that we cannot connect to via Admin API.
+   - `WAIT_FOR_LOG_RECOVERY` -  Nodes that have `RECOVERING` state.
+   - `RESTART` - Nodes that have non-empty list of reasons from the predicate function and have not been restarted yet (Context.numRestarts == 0).
+   - `MAYBE_RECONFIGURE` - Broker nodes (including combined nodes) that have an empty list of reasons and not been reconfigured yet (Context.numReconfigs == 0).
+   - `NOP` - Nodes that have been restarted or reconfigured at least once (Context.numRestarts > 0 || Context.numReconfigs > 0 ) and have either
+             `LEADING_ALL_PREFERRED` or `SERVING` state.
+
+4. Wait for nodes in `WAIT_FOR_LOG_RECOVERY` group to finish performing log recovery. 
+   - Wait for each node to have `SERVING` within the `postOperationalTimeoutMs`. 
+   - If the timeout is reached for a node and its `numAttempts` is greater than or equal to `maxAttempts`, throw `UnrestartableNodesException` with the log recovery progress (number of remaining logs and segments). Otherwise increment node's `numAttempts` and restart from step 3.
+
+5. Restart nodes in `RESTART_FIRST` category either one by one in the following order unless all nodes are combined
+and are in `NOT_RUNNING` state:
+   - Pure controller nodes
+   - Combined nodes
+   - Broker only nodes
+
+   If all controllers are combined and are in `NOT_RUNNING` state, restart all nodes in parallel and wait for them to have `SERVING`. Explained more in detail below.
```

Comment:
I think this would be easier to read if you used a sublist for the two cases (controller+not running vs otherwise)